src/shell/xdg/frame/mod.rs

```diff
@@ -0,0 +1,133 @@
+//! The frame to use with XDG shell window.
+
+use crate::reexports::client::protocol::wl_surface::WlSurface;
+use crate::reexports::protocols::xdg::shell::client::xdg_toplevel::ResizeEdge;
+
+use crate::shell::xdg::window::{WindowManagerCapabilities, WindowState};
+
+pub mod fallback_frame;
+
+/// The interface for the client side decorations.
+pub trait DecorationsFrame: Sized {
+    /// Emulate click on the decorations.
+    ///
+    /// The `click` is a variant of click to use, see [`FrameClick`] for more information.
+    ///
+    /// The return value is a [`FrameAction`] you should apply, this action could be
+    /// ignored.
+    ///
+    /// The location of the click is the one passed to [`Self::click_point_moved`].
+    fn on_click(&mut self, click: FrameClick, pressed: bool) -> Option<FrameAction>;
+
+    /// Emulate pointer moved event on the decorations frame.
+    ///
+    /// The `x` and `y` are location in the surface local coordinates relative to the `surface`.
+    ///
+    /// The return value is the new cursor icon you should apply to provide better visual
+    /// feedback for the user. However, you might want to ignore it, if you're using touch events
+    /// to drive the movements.
+    fn click_point_moved(&mut self, surface: &WlSurface, x: f64, y: f64) -> Option<&str>;
+
+    /// All clicks left the decorations.
+    ///
+    /// This function should be called when input leaves the decorations.
+    fn click_point_left(&mut self);
+
+    /// Update the state of the frame.
+    ///
+    /// The state is usually obtained from the [`WindowConfigure`] event.
+    ///
+    /// [`WindowConfigure`]: crate::shell::window::WindowConfigure
+    fn update_state(&mut self, state: WindowState);
+
+    /// Update the window manager capabilites.
+    ///
+    /// The capabilites are usually obtained from the [`WindowConfigure`] event.
+    ///
+    /// [`WindowConfigure`]: crate::shell::window::WindowConfigure
+    fn update_wm_capabilities(&mut self, wm_capabilities: WindowManagerCapabilities);
+
+    /// Resize the window to the new size.
+    ///
+    /// The size must be without the borders, as in [`Self::subtract_borders]` were used on it.
+    ///
+    /// **Note:** The [`Self::update_state`] and [`Self::update_wm_capabilities`] **must be**
+    /// applied before calling this function.
+    ///
+    /// # Panics
+    ///
+    /// Panics when resizing the hidden frame.
+    fn resize(&mut self, width: u32, height: u32);
```

Comment:
What should occur if the caller passes 0 for both the width and height

Reply:
That's a good question, do we want `NonZero`?

Reply:
I think `NonZero` would be good here.

Reply:
I'd suggest for configure to pass `NonZeroU32` as well then, because it's actually a `NonZeroU32` already.

Reply:
https://wayland.app/protocols/xdg-shell#xdg_toplevel:event:configure

I think a configure event may set width or height to 0 if the client should "decide its own window dimenstions"

Reply:
I've updated to use `NonZeroU32`.